abi/derive/src/lib.rs

```diff
@@ -0,0 +1,33 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+// This file is part of ink!.
+//
+// ink! is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// ink! is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+extern crate proc_macro;
+
+#[macro_use]
+mod error;
+
+mod has_layout;
+mod impl_wrapper;
+
+use proc_macro::TokenStream;
+
+#[proc_macro_derive(HasLayout)]
```

Comment:
Would be good to see this exercised in a test outside of `lang/gen`, at least the happy path

Reply:
Can we do this in a follow-up PR? I really just want to get this over the finishing line.